### PR TITLE
Add CDash URLs to builds and nonpassing tests with same base repo version 

### PR DESCRIPTION
**Pull request by bartlettroscoe**
This PR add CDash URLs getting printed and written to file for the builds and non-passing tests on CDash for the same base repo Git version SHA1 (see Issue #483).  For more info, see the new section [Links to results on CDash](https://github.com/TriBITSPub/TriBITS/blob/483-cdash-revision-urls/tribits/ctest_driver/TribitsCTestDriverCore.cmake#L1170).

## Notes to Reviewers

* This PR looks larger than it actually is because I moved code around between files and created new files.  And a lot of the lines added are for tests and copyright/license headers
* It is likely better to review this PR commit-by-commit and by ignoring whitespace
* The only larger commit is [Produce CDash URLs to all builds and nonpassing tests for same repo version (#483)](https://github.com/TriBITSPub/TriBITS/pull/485/commits/1738ccdd861e0da6bfaa683d4ae5eeddea643971).

### CODE REVIEW

**Review by keitat — APPROVED** [review]
LGTM

**Review by rabartlett1972 — APPROVED** [review]
Approving so the merge can happen but it would be good to get a post-merge review from @KyleFromKitware

File: test/ctest_driver/TribitsExampleMetaProject/CMakeLists.txt
Comment (KyleFromKitware) [inline test/ctest_driver/TribitsExampleMetaProject/CMakeLists.txt:101]: Shouldn't be indented

File: tribits/ctest_driver/TribitsReadTagFile.cmake
Comment (bartlettroscoe) [reply tribits/ctest_driver/TribitsReadTagFile.cmake:13]: Thanks for the tip.  I will make the change.  I did not realize that a CMake function did not set a new policy scope like setting up a new variable scope.  This is the exact opposite of the `include()` command.  Including a file with `include()` sets a new policy scope but does not set a new variable scope.  CMake never ceases to surprise me.

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (bartlettroscoe) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:77]: Can you use `Git::Git` in the `COMMAND` for `execute_process()`?  There is no mention that you can in the documentation:

* https://cmake.org/cmake/help/v3.23/command/execute_process.html

File: tribits/ctest_driver/TribitsGetCDashUrlsInsideCTestS.cmake
Comment (bartlettroscoe) [reply tribits/ctest_driver/TribitsGetCDashUrlsInsideCTestS.cmake:431]: Fixed in a later commit 1738ccdd861e0da6bfaa683d4ae5eeddea643971.

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (bartlettroscoe) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:86]: Thanks!  This was a copy and paste from the function `tribits_generate_single_repo_version_string()` that actually needs those quotes because there are spaces in the `--pretty=format:<value>`.  (In that case `--pretty=format:"%h [%ad] <%ae>"`).  I verified that I can remove the spaces (and there is a strong test that verifies the code still works as it should).

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (bartlettroscoe) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:77]: The advantage of using the variable `GIT_EXECUTABLE` instead of the target `Git::Git`` is that I can test that function in a `ctest -P` unit test script.  You can't define targets in a ctest -P script.

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (KyleFromKitware) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:86]: I see.

CMake has three types of arguments: unquoted, quoted, and bracketed. If the argument starts with `"`, it's a quoted argument. If not, it's an unquoted argument, even if it contains a quote later on. So, for `tribits_generate_single_repo_version_string()`, you could change it to `"--pretty=format:%h [%ad] <%ae>"` and not have to strip off the quotes afterwards.

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (KyleFromKitware) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:77]: OK, so this is in a `-P` script. In that case using `GIT_EXECUTABLE` would be best. You may still want to check out `FindGit` though.

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (bartlettroscoe) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:86]: I am pretty sure I tried that and it did not work.  I can try it again.

Comment (KyleFromKitware) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:86]: I just tried the following locally and it worked fine:

```cmake
execute_process(COMMAND git log -1 "--pretty=format:%h [%ad] <%ae>")
```

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (bartlettroscoe) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:77]: This module can be used in a cmake -P script and the unit test driver that runs some unit tests on this fucntion is run with cmake -P.

What does `find_package()` do when called from a cmake -P script where targets are not allowed?

Comment (KyleFromKitware) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:77]: It is legal to call `find_package()` from inside `cmake -P` as long as the package doesn't try to create a target. [`FindGit`](https://cmake.org/cmake/help/latest/module/FindGit.html) specifically has a provision to not attempt to create the `Git::Git` target if it's called from inside `cmake -P`:

> *New in version 3.14:* The module defines the following `IMPORTED` targets (when [`CMAKE_ROLE`](https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_ROLE.html#prop_gbl:CMAKE_ROLE) is `PROJECT`):
>
> `Git::Git`
>
> Executable of the Git command-line client.

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (bartlettroscoe) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:86]: Okay, that does seem to work correctly.  There are strong tests that break of you don't produce this line correctly.  I will include this change in the follow up PR.

File: test/core/TestingFunctionMacro_UnitTests.cmake
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:497]: I have switched this over to `find_package(Git REQURIED)` and it does work when called in a cmake -P script (which is how the unit tests in this file are executed).

I will include this in the follow up PR.

NOTE: All other cases where TriBITS was finding Git was already using `find_package(Git)` (even in a `ctest -S` script) so this is great!

File: tribits/ctest_driver/TribitsReadTagFile.cmake
Comment (bartlettroscoe) [reply tribits/ctest_driver/TribitsReadTagFile.cmake:13]: So, just to be clear, setting `cmake_policy()` at the file-level scope will impact the policies for all of the functions and macros defined in that file and in all files pulled in from this file with `include()` and `add_subdirectory()`?

Then what is the use case for allowing a CMake function to call `cmake_policy()` and impact the policy of the code in another file?  Should that not be allowed or at least warned against by CMake?

Now I can understand a macro calling  `cmake_policy()`  and impacting the policy of the calling code but it makes no sense for a function to do so, IMHO.

Comment (KyleFromKitware) [reply tribits/ctest_driver/TribitsReadTagFile.cmake:13]: > Then what is the use case for allowing a CMake function to call cmake_policy() and impact the policy of the code in another file?

It's supposed to be defined by the top-level `CMakeLists.txt`, where it propagates to the rest of the project by `include()` and `add_subdirectory()`. If a file needs a policy set to a specific setting, it should set it itself with `cmake_policy(SET)`.

Comment (bartlettroscoe) [reply tribits/ctest_driver/TribitsReadTagFile.cmake:13]: Right, but why even allow a function() to call `cmake_policy(SET ...)`?  Should CMake at least warn against this?

File: tribits/core/utils/TribitsParseArgumentsHelpers.cmake
Comment (bartlettroscoe) [reply tribits/core/utils/TribitsParseArgumentsHelpers.cmake:87]: Addressed in commit https://github.com/TriBITSPub/TriBITS/pull/486/commits/165afcad343c2906b5db16a8ed09cc260e13e96a in new PR #486

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (bartlettroscoe) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:77]: We will keep using `GIT_EXECUTABLE` instead of `Git::Git` since this is called in cmake -P scripts as well.

File: test/core/TestingFunctionMacro_UnitTests.cmake
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:497]: Addressed in commit https://github.com/TriBITSPub/TriBITS/pull/486/commits/6afc55f4006cdd9f09bae544e09cd95ac7cbdeaa in new PR #486.

File: tribits/ctest_driver/TribitsReadTagFile.cmake
Comment (bartlettroscoe) [reply tribits/ctest_driver/TribitsReadTagFile.cmake:13]: Addressed in commit https://github.com/TriBITSPub/TriBITS/pull/486/commits/040023215978c9d39735c8200b46510504361ccc in new PR #486

File: tribits/ctest_driver/TribitsGetCDashUrlsInsideCTestS.cmake
Comment (bartlettroscoe) [reply tribits/ctest_driver/TribitsGetCDashUrlsInsideCTestS.cmake:461]: Addressed in commit https://github.com/TriBITSPub/TriBITS/pull/486/commits/165afcad343c2906b5db16a8ed09cc260e13e96a in new PR #486

File: tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake
Comment (bartlettroscoe) [reply tribits/core/package_arch/TribitsGitRepoVersionInfo.cmake:86]: Addressed in commit https://github.com/TriBITSPub/TriBITS/pull/486/commits/4bd215a374ce28779416e62f85c7b9f9ec9976f8 in new PR #486

File: tribits/ctest_driver/TribitsReadTagFile.cmake
Comment (KyleFromKitware) [reply tribits/ctest_driver/TribitsReadTagFile.cmake:13]: The purpose of the function could be to set policies. I think it's just a matter of making sure you understand policy scope.

**Pull request merged**
